ops/ruby/devops-toolkit/tasks/monitoring_assets.rb

```diff
@@ -1,65 +1,227 @@
-      @logger.info("found #{policies.count} alerting policies")
+      policies = @alert_client.list_alert_policies(@project_path)
+      logger.info("found #{policies.count} alerting policies")
       counts['policies'] = policies.count
 
       policies.each do |policy|
-        @logger.info("\t#{policy.display_name}, #{policy.name}, #{policy.conditions.count} conditions")
+        logger.info("\t#{policy.display_name}, #{policy.name}, #{policy.conditions.count} conditions")
       end
 
-      @logger.info("Total counts for #{env.short_name}: #{counts}")
+      logger.info("Total counts for #{@current_env.short_name}: #{counts}")
+    end
+  end
+
+  DO_BACKUP_ALL = false # TODO(jaycarlton): make this an option
+
+  # For every environment in the envs_file, fetch all items of interest from the appropriate
+  # Stackdriver APIs. (Note that in future we may wish to broaden this to other )
+  def backup_config
+    visit_envs do |env_bundle|
+      backup_alert_policies
+      if DO_BACKUP_ALL
+        backup_all_metrics
+      else
+        backup_custom_metrics
+        backup_user_logs_based_metrics
+      end
+      backup_dashboards
+      backup_group_members
+      backup_monitored_resources
+      backup_notification_channels
+      backup_notification_channels
     end
+  end
+
+  private
+
+  def backup_alert_policies
+    policies = @alert_client.list_alert_policies(@project_path)
+    backup_assets(policies)
+  end
+
+  def backup_custom_metrics
+    custom_metrics = @metric_client.list_metric_descriptors(@project_path, {filter: CUSTOM_METRIC_FILTER})
+    backup_assets(custom_metrics)
+  end
+
+  def backup_dashboards
+    dashboards = @dashboard_client.list_dashboards(@current_env.formatted_project_number)
+    backup_assets(dashboards)
+  end
+  def backup_user_logs_based_metrics
+    user_logs_based_metrics = @metric_client.list_metric_descriptors(@project_path, {filter: USER_LOGS_BASED_METRIC_FILTER})
+    backup_assets(user_logs_based_metrics)
+  end
+
+  def backup_monitored_resources
+    resources = @metric_client.list_monitored_resource_descriptors(@project_path)
+    backup_assets(resources)
+  end
+
+  # Grab all the metrics, including very many defined by Google and partners. This method is here
+  # more for completeness and to allow introspection/navigation of them than anything else, as we
+  # can't write to these to restore them.
+  def backup_all_metrics
+    metrics = @metric_client.list_metric_descriptors(@project_path)
+    backup_assets(metrics)
+  end
+
+  def backup_group_members
+    members = @group_client.list_groups(@project_path)
+    backup_assets(members)
+  end
+
+  def backup_notification_channels
+    channels = @notification_channel_client.list_notification_channels(@project_path)
+    backup_assets(channels)
+  end
+
+  # Backup a collection of Stackdriver assets
+  def backup_assets(assets)
+    assets.each(&method(:backup_asset))
```

Comment:
Interesting, first time seeing this syntax

Reply:
Yeah the IDE did that for me on a refactor.